src/libmain/progress-bar.cc

```diff
@@ -71,53 +144,220 @@ class ProgressBar : public Logger
 
         bool active = true;
         bool haveUpdate = true;
+
+        std::map<LineId, std::string> statusLines;
+
+        /* How many lines need to be erased when redrawing. */
+        size_t prevStatusLines = 0;
+
+        bool helpShown = false;
+
+        std::optional<std::promise<std::optional<char>>> prompt;
     };
 
+    const bool isTTY;
+
     Sync<State> state_;
 
     std::thread updateThread;
+    std::thread inputThread;
 
     std::condition_variable quitCV, updateCV;
 
-    bool printBuildLogs;
-    bool isTTY;
+    std::optional<struct termios> savedTermAttrs;
+
+    Pipe inputPipe;
```

Comment:
`inputPipe` is only used to cancel the `inputThread`. Is that right? It didn't occur to me in the first passage of reading the thread that was the intend of it. It is a bit unfortunate to have the whole indirection through the pollfd.